test/__testutils__/mockPrismicMigrationAPI.ts

```diff
@@ -0,0 +1,159 @@
+import type { TestContext } from "vitest"
+
+import type { RestRequest } from "msw"
+import { rest } from "msw"
+
+import type { PrismicDocument, WriteClient } from "../../src"
+import type {
+	PostDocumentParams,
+	PostDocumentResult,
+	PutDocumentParams,
+	PutDocumentResult,
+} from "../../src/types/api/migration/document"
+
+type MockPrismicMigrationAPIArgs = {
+	ctx: TestContext
+	client: WriteClient
+	writeToken?: string
+	migrationAPIKey?: string
+	requiredHeaders?: Record<string, string>
+	existingDocuments?: (PostDocumentResult | PrismicDocument)[] | number
+	newDocuments?: { id: string; masterLanguageDocumentID?: string }[]
+}
+
+type MockPrismicMigrationAPIReturnType = {
+	documentsDatabase: Record<
+		string,
+		PostDocumentResult & Pick<PrismicDocument, "data">
+	>
+}
+
+export const mockPrismicMigrationAPI = (
+	args: MockPrismicMigrationAPIArgs,
+): MockPrismicMigrationAPIReturnType => {
+	const repositoryName = args.client.repositoryName
+	const migrationAPIEndpoint = args.client.migrationAPIEndpoint
+	const writeToken = args.writeToken || args.client.writeToken
+	const migrationAPIKey = args.migrationAPIKey || args.client.migrationAPIKey
+
+	const documentsDatabase: Record<
+		string,
+		PostDocumentResult & Pick<PrismicDocument, "data">
+	> = {}
+
+	const validateHeaders = (req: RestRequest) => {
+		if (args.requiredHeaders) {
+			for (const name in args.requiredHeaders) {
+				const requiredValue = args.requiredHeaders[name]
+
+				args.ctx.expect(req.headers.get(name)).toBe(requiredValue)
+			}
+		}
+	}
+
+	if (args.existingDocuments) {
+		if (typeof args.existingDocuments === "number") {
+			for (let i = 0; i < args.existingDocuments; i++) {
+				const document = args.ctx.mock.value.document()
+				documentsDatabase[document.id] = {
+					title: args.ctx.mock.value.keyText({ state: "filled" }),
+					id: document.id,
+					lang: document.lang,
+					type: document.type,
+					uid: document.uid,
+					data: document.data,
+				}
+			}
+		} else {
+			for (const document of args.existingDocuments) {
+				if ("title" in document) {
+					documentsDatabase[document.id] = { ...document, data: {} }
+				} else {
+					documentsDatabase[document.id] = {
+						title: args.ctx.mock.value.keyText({ state: "filled" }),
+						id: document.id,
+						lang: document.lang,
+						type: document.type,
+						uid: document.uid,
+						data: document.data,
+					}
+				}
+			}
+		}
+	}
+
+	args.ctx.server.use(
+		rest.post(`${migrationAPIEndpoint}documents`, async (req, res, ctx) => {
+			if (
+				req.headers.get("authorization") !== `Bearer ${writeToken}` ||
+				req.headers.get("x-api-key") !== migrationAPIKey ||
+				req.headers.get("repository") !== repositoryName
+			) {
+				return res(ctx.status(403), ctx.json({ Message: "forbidden" }))
+			}
+
+			validateHeaders(req)
+
+			const body = await req.json<PostDocumentParams>()
+
+			const newDocument = args.newDocuments?.shift()
```

Comment:
Sure, it's a way to have predictable IDs for new document IDs when testing for it and to test for expected master language document IDs (because those aren't returned by the API in its response)

Reply:
Add a short comment about it